### DISCUSSION TITLE
Stopped Emitter Leaving Particles Behind

### DISCUSSION THREAD

**Discussion by ben-randle-core** [creativelifeform/three-nebula#147] [General]
I'm experiencing a bit of a problem when I start a system of emitters then stop those emitters. After `stopEmit` has been called on the emitters I seem to be left with some particles that are affecting the appearance of the scene. I've written the following function for toggling emitters within a created system on/off:

```
toggleEmitters(emit) {
        this._system.emitters.forEach(emitter => {
            if (emit) {
                emitter.emit();
            }
            else {
                emitter.stopEmit();
            }
        });
    }
```

After calling `toggleEmitters(false);` I get the following in my scene. The emitter I am toggling is a sort of flame/fire emitting from the area that says "WILD". The majority of these particles move upwards and you can see from the screenshot below that the "K" above has been "bleached" by the particle emitter that has actually already been stopped (and is stopped when this screenshot is taken). The "K" below still has it's green appearance because the particles have not travelled that direction.

![image](https://user-images.githubusercontent.com/48215725/112137985-de0c4c00-8bc8-11eb-88c5-9d577b065ebb.png)

I would appreciate any help with this issue. I've tried manually calling `emitter.removeAllParticles();` too which didn't seem to have any affect.

Thanks in advance,
Ben

**Comment by rohan-deshpande** [comment]
When you add the particles, try adding them to a THREE.Group that is already in the Scene and perhaps cleaning up all children of that particular Group when you want the particles to all get deleted?

**Reply by ben-randle-core** [reply]
Hi Rohan, thanks for the response.

I've been messing around with this for a while now and I still haven't been able to get around the problem. It looks to me like the crux of the problem is that when `stopEmit` is called it sets `isEmitting` to false, then in the `update` loop it stops doing anything with the particles. This means that existing particles in the emitter aren't allowed to live out their lifetimes and die naturally. So they are just left lying around indefinitely.

I thought I might get around this by doing the following:

```
emitter.stopEmit();
emitter.particles.forEach(p => p.destroy());
```

I had hoped that the above would just get rid of the particles immediately (which isn't ideal but acceptable) but this also left the particles hanging around in the scene. I found this very odd because the `destroy` method sets `this.parent = null;` according to the documentation!

Thanks for the help,
Ben

**Reply by ben-randle-core** [reply]
I've just tried emulating what the `update` loop would be doing if it were getting past the `this.isEmitting` check and it has removed them immediately (as expected).

```
                emitter.stopEmit();
                emitter.particles.forEach(particle => {
                    emitter.parent && emitter.parent.dispatch("PARTICLE_DEAD", particle);
                    emitter.bindEmitterEvent && emitter.dispatch("PARTICLE_DEAD", particle);
                    emitter.parent.pool.expire(particle.reset());
                });
                emitter.particles.length = 0;
```

So this is a work around I could use but still isn't ideal. As I said in my previous response I would expect the particles that have already been spawned when `stopEmit` is called to naturally live out their lifetime until their death rather than just stopping and lingering indefinitely. I think this is probably a bug.

Thanks,
Ben

**Reply by rohan-deshpande** [reply]
Ok, thanks for this. I agree that the expected behaviour should probably be that the emitter stops emitting but the existing particles continue to live until dead. So yeah, this is a bug. I'll make an issue and get to it shortly.

**Comment by ben-randle-core** [comment]
Any movement on this @rohan-deshpande ?

**Reply by rohan-deshpande** [reply]
I'll try to get to it this week, sorry for the delay

**Reply by rohan-deshpande** [reply]
Sorry for the massive delay here, you might have moved on from the library but just wanted to say I'll definitely investigate this now. I'm not having the same problem in my apps though. If you could post a codesandbox to reproduce the issue it would make it easier for me to solve.

**Comment by wmcmurray** [comment]
Another workaround for manually removing particles from the scene 💩 

```js
let particle;
while (emitter.particles.length) {
  particle = emitter.particles.pop();
  particle.target.removeFromParent();
  particle.destroy();
}
```

**Comment by Elmanzalawy** [comment]
I had a similar issue with emitters leaving static particles on calling  the `stopEmit()` method.
it worked for me by changing 
```js
emitter.stopEmit()
```
to
```js
emitter.totalEmitTimes = -1
```
